### DISCUSSION TITLE
set timer between messages within a loop

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#285] [Q&A] [answered]
Hi everybody,

I struggle to have a timer running properly in a loop.

For a "bulk memory" request I need to set a 200ms interval between sent messages.
the messages are sysex requests for individual program from #1 to #100 

`0xF0, 0x43, 0x20, 0x7E, 0x4C, 0x4D, 0x20, 0x20, 0x38, 0x41, 0x32, 0x31, 0x4D, PROGRAM, 0xF7`

so let's have a loop to send the 100 requests without delay between messages:

```
for i = 1,100 do
panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x43, 0x20, 0x7E, 0x4C, 0x4D, 0x20, 0x20, 0x38, 0x41, 0x32, 0x31, 0x4D, i, 0xF7})) 
end
```

Now I need to set an interval of time of at least 200ms between requests:

```
for i = 1,100 do
	panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x43, 0x20, 0x7E, 0x4C, 0x4D, 0x20, 0x20, 0x38, 0x41, 0x32, 0x31, 0x4D, i, 0xF7}))
	timer:setCallback(2, TimerCallback) -- TimerId 2
	timer:startTimer(2, 200) -- Timer Id 2 200ms

-- there should be 200ms before the loop cycles back but it doesn't :( 
end
```

My timerCallback function is : 

```
TimerCallback = function(timerId)

	if timerId == 2 then
		timer:stopTimer(timerId)
	end

end
```

This should set an delay of 200ms between each iteration of the loop but I have still the 100 messages sent straight away then the delay occures.

Do you know how I can manage to have this working?

Thanks a lot in advance for your help. This timer function is a bit tricky sometimes.

Damien

**Comment by dobo365** [comment, marked as answer]
Hi Damien,
  thx for this question that allows me to explain you what a timer is and how it is working ;-)
I think I already did that in the old forum but it is good to repeat.

A timer is itself a loop running continuously at a certain timing until you stop it.
So, you don't need to have your 1-100 for loop there and starting/stopping the same timer ID 2 100 times.

From my _ever expanding never published_ Step by Step guide 2.x (will further work on it during summer so I can publish it):

**The philosophy of a timer is that it loops by executing its code at a specific time interval until it is stopped.**
Juce documentation: ((https://www.juce.com/doc/classTimer#details ))

Some properties:
- Timers are identified by a unique ID
- Before starting a timer with the timer:startTimer(ID, time_interval) function, it is needed to indicate the function that will be called (the callback function) with the timer:setCallback (ID, callback_function)
- All timers can be in the same callback function provided you use the timer ID as argument and use an if..then..else statement to differentiate the timer’s actions
- A timer must be stopped and this is done with the timer:stopTimer(ID) function
- The time interval isn't guaranteed to be precise to any more than maybe 10-20ms, and the intervals may end up being much longer than requested if the system is busy. Because the callbacks are made by the main message thread, anything that blocks the message queue for a period of time will also prevent any timers from running until it can carry on.

So:
- In your main code. You should only initialize a counter and start the timer
```
iCounter = 0
timer:setCallback(2, TimerCallback) -- TimerId 2
timer:startTimer(2, 200) -- Timer Id 2 200ms
```
- Then you let the timer run in the timer callback method until it reaches 100 and you stop it there
```
TimerCallback = function(timerId)

	if timerId == 2 then
		if iCounter < 100 then
			panel:sendMidiMessageNow(CtrlrMidiMessage({0xF0, 0x43, 0x20, 0x7E, 0x4C, 0x4D, 0x20, 0x20, 0x38, 0x41, 0x32, 0x31, 0x4D, iCounter, 0xF7}))
			iCounter = iCounter + 1
		else
			timer:stopTimer(timerId)
		end
	end

end
```

I didn't verify the code above but you should get the philosophy.
You may need to put 101 in the if condition.
Good luck!

**Reply by damiensellier** [reply]
Hi Dominique,
Thanks a lot for the explanation. I really take time to read JUCE documentation everytime I come across an issue.
It's very detailed but lacks of exemple like there's some available in the LUA docs.

I totally understood how it works thanks to your reply. I was making a loop starting another loop so it was not working as intended :) 

I had it working temporarily with os.execute(sleep(200)) but want to avoid it at all costs because it hangs the system for a while. I know that if I had to use os.execute(sleep(T)) there's something wrong in my method.

With your help it's working like a charm. Thank you so much professor. Good luck for the  "Step by Step guide 2" the first one really helped me when I started and I understood everything about modulators/components.

I am very happy to flag this post as solved. Have a nice day!

Damien

**Reply by DRBAUS** [reply]
Okay, I understand this and in the MIDI Monitor I see how it works. If I want it sends the message each 500ms. Cool! 

What I am looking for now is a way to send a different sysex string everytime the loop is running.

I was able to make a working script (using Timers) which sends 15 codes one after the other by using a Timer for each code. But 15 scripts each containing 15 Timers, times 3 per Layer....I already lost count on the Timers...lol 

There must be an easier and better way that I don't see.

**Reply by dobo365** [reply]
Hallo!
What you could do is to only use one timer with your 500ms.
Before starting the timer, set a variable = 1 for example
In the timer, build an IF..THEN..ELSEIF..ELSEIF...ELSEIF...END structure
In the first IF, test for the variable=1, set the same variable to 2, send your first message
In the second IF, test for the variable=2, set the same variable to 3, send your second message
In the third IF, test for the variable=3, set the same variable to 4, send your third message
and so on
When you are at 15, stop the timer
I think this should work fine to send your 15 messages one by one at 500ms interval
Best...Dominique

**Reply by DRBAUS** [reply]
Bonjour Goodweather!

Ahaa...so THAT's how it works. Got it!

On the last ELSEIF I also put a counter = 1 so it will reset. The next time a double click is received the
script will run again from the beginning.

This is what I was looking for.

Merci beaucoup,

Robin.

**Comment by dnaldoog** [comment]
This is a great explanation !

Every time I have to do a timer I have to think about the logic so this really helps.

**Comment by Ted-juh** [comment]
Case: In a CustomComponent, the UiCustomDoubleClickCallback always fires a customMouseDownCallback as well. But because of some weird obscure reason the second mouse down seems to trigger "after" the double click. 

If double clickitty click:
-- triggers single click
-- triggers double click
-- triggers single click again.

I use a function to paint something with a single click. Want to delete it with a double click. Guess you can figure out what happens. Going over the Juce forums, this seems to be the order for the mouse clicks. Jules advises using "getNumberOfClicks" or to use a timer. 

-- getNumberOfClicks() -- just counts the amount of clicks. Even a click after a minute still adds up. So not usable right away.
-- timer -- Although Goodweather provided a good explanation a few posts ago, I'm still struggling with it.  If I put a timer in a mouseDownCallback, the timer resets every time the button is pressed. 

So the question is: How do I count the number of clicks in a set period of milliseconds?
-- mouseDown --> wait or set timer for 20 ms and see if another click happens --> outcome should be 0 (single click) or > 0 (double click)

**Comment by dnaldoog** [comment]
No idea about this! Have you checked all the possible JUCE mouse methods?
Sure you have and maybe not all functionality is available in Ctrlr.
https://stackoverflow.com/questions/5965197/intercepting-mouse-events-on-filebrowsercomponent-in-juce-library
You would know way more than me about such things.

Actually for my Rev 500 panel - do you know how to change the font family
inside a uiComponent - I want to use digital-7.

Also as far as I can tell there is no information about the data in the
manual - it just says 66 total bytes etc, but no assignations so I might
have to go through and do byte changes comparisons by changing each value
on the unit. That could be very tedious. Or just ignore all that and write
a JSON file saving thing that can be saved or loaded and then transmitted.

Each control is using CC but they have this interpolate thing which I
encountered in the C-PLG-150AN where a button 0-127 has 200 values, or was
it the other way round. Other buttons are the same - for example PRE-DELAY
seems to have 200 values on the machine but CC is only 0-127 according to
the manual.

You can see it on the EFF LEVEL bar 0-127 goes from 0-20 bars
This is the function

function interpolate(oldValue)
    --The PLG-150 AN editor somehow divides 200 values between 128 sysex
messages
    local new_value=0
    local old_value=oldValue
    local old_min=0
    local old_max=127
    local new_min=0
    local new_max=20
    new_value=((old_value-old_min)/(old_max-old_min))*(new_max-new_min)+
new_min
return math.ceil(new_value)
end --end function
---------------------------------------------------------

Attached is far from working panel.

I got part way through working on a JSON file that I didn't post on the
forums, but I notice the subject has come up on the forum again. I had it
partly working well but not perfectly. Has a lot of potential. Like I said
in that post I have just written lua tables to text files - that works very
well too.

**Reply by Ted-juh** [reply]
Import a font that you want to use to your resources or find the path where the font is located. Loads of free fonts on the internet that are close to that digital-7 font, so I guess that just importing from resources is the easy way to do it. Add the following lines to your paint function of a custom component:

newFont = resources:getResourceAsFont("getdigital")
g:setFont(newFont)
g:setFont(10.0)

getdigital = the name of the font without the extension. 

If you want to add it to a non-custom component you probably have to override L&F.

**Reply by Ted-juh** [reply]
I'll get back at you about the sysex for the Rev500. Took a quick look when you sent me the PDF the other day. Noticed an asterix *2 that you had to ask the Yamaha helpdesk for something. I couldn't figure out if it was about the parameters you are talking about. So, patience :).

**Reply by damiensellier** [reply]
```
Also as far as I can tell there is no information about the data in the
manual - it just says 66 total bytes etc, but no assignations so I might
have to go through and do byte changes comparisons by changing each value
on the unit. That could be very tedious. Or just ignore all that and write
a JSON file saving thing that can be saved or loaded and then transmitted.

Each control is using CC but they have this interpolate thing which I
encountered in the C-PLG-150AN where a button 0-127 has 200 values, or was
it the other way round. Other buttons are the same - for example PRE-DELAY
seems to have 200 values on the machine but CC is only 0-127 according to
the manual.
```

@dnaldoog This was the exact same thing for the SPX990 I worked on.
You have to decode every single parameter value from the Current Buffer Bulk Dump. And update those parameters via sysex or midi CC. What's cool in the Rev500 is that yamaha added acces to those sysex messages. It was not available for the old SPX.  No Buffer Was avalaible nor the sysex parameter value change. 

But what's each actual parameter value when you recall a program? This was my first dilema. I needed the pannel to reflect what's the current Program different values. Yamaha is not helping with that.
The only source is the Program dump you get with Current Buffer Bulk Dump Request. (P26 #5 of the REV500 product manual.)
One by one, you have to modify a parameter on the machine (min to max), then request a buffer dump and check what's moving on the sysex message. Some are linear values, but most of the parameters in yamaha FX use fixedvalues. 
I made a chart for the SPX with all the parameters for every Algorithm and their respective byte position from the Bulk Dump.
In this chart I have the min value, max value, if it's linear (0 to 1000 with 0.1 increments for exemple) or if it's arbitrary for fixed sliders (0.1s 0.2ms 0.3ms 0.4ms 0.5ms 1.ms 1.5ms 2ms 3ms 4ms 5ms 10ms etc).

Just that took me like 10 days. Just because I needed the damn current value of each parameter to start with :)

Then once you have all your parameters details you can start updating the values to a 0-127 midi CC. In my panel I update the midi CC value with a script "when mod value changes".


1. check which algo is used in the current program
2. depending on the algo, update all the sliders of the panel with the proper name and value properties (min max etc)

for exemple this is what I have if the algo used is "Reverb" for the 1st paramter : 

```
	MainFxParamValue01 = SysexDecodedMessage:getByte(17)
--console("Displays MainFxParamValue01 : "..(MainFxParamValue01))

	MainFxParam01 = panel:getModulatorByName("MainFxParam01")
	MainFxParam01:getComponent():setProperty("componentVisibility", 1, false)
	MainFxParam01:getComponent():setProperty("componentVisibleName", "RevTyp", true)
	MainFxParam01:getComponent():setProperty("uiSliderMin", 0, true)
	MainFxParam01:getComponent():setProperty("uiSliderMax", 3, true)
	MainFxParam01:getComponent():setProperty("uiSliderDoubleClickEnabled", 1, true)
	MainFxParam01:getComponent():setProperty("uiSliderDoubleClickValue", 0, true)
	MainFxParam01:getComponent():setProperty("uiFixedSliderContent","Hall\nRoom\nVocal\nPlate", true)
	MainFxParam01:getComponent():setValue(MainFxParamValue01, true)
	MainFxParam01:setProperty("modulatorCustomIndex","02", true)
	MainFxParam01:setProperty("modulatorCustomName","1", true)
	MainFxParam01:setProperty("modulatorCustomIndexGroup","", true)
	MainFxParam01:setProperty("modulatorCustomNameGroup","1", true)
	MainFxParam01:getComponent():setProperty("uiType","uiFixedImageSlider", true)
```

And the script for luaModulatorValueChange : 


```
function CCValueSend(mod, value, source)

modValueMin = mod:getComponent():getPropertyInt("uiSliderMin")
modValueMax = mod:getComponent():getPropertyInt("uiSliderMax")
local ModValue = mod:getModulatorValue()
	
	ccValueStep = 127 / (modValueMax + math.abs(modValueMin))
	--console("Displays ccValueStep : "..(ccValueStep))

	if modValueMin < 0 then
		ccValueRel = math.abs(modValueMin) + value
	else
		ccValueRel = value
	end
	--console("Displays ccValueRel : "..(ccValueRel))

	CcValue =  ccValueStep * ccValueRel
	--console("Displays CcValue : "..(CcValue))

	panel:sendMidiMessageNow(CtrlrMidiMessage({0xB0, 0x12, CcValue})) -- Default CC control #18 is 0x12  for the SPX990

end -- end f
```
Yes you have to scale down your values to 0-127 for the CC. Yamaha always ask 0-127 even if it's OFF/ON(0 or 127) or 1, 2 , 3 (0-64-127)
Since there's a Buffer Edit dump request available in the REV500, you can request it everytime you click on a modulator and play ping pong with the machine to have all parameters synced up from the panel and the machine.

**Reply by dnaldoog** [reply]
Well I thought the message I replied to via email to Tedjuh was private, hence the question about importing fonts to a _uiComponent_ which is of course off topic, but it ended up getting published on the site! I had no idea. So the original topic of this thread has taken a new turn.

It's almost coincidental that you were working on the SPX90, and I knew that, but what I didn't know is that it too has this same 0-127 interpolation system! I will look at your way of doing it. The code I posted was working for the [C-PLG150](https://ctrlr.org/c-plg150-an/) panel, so it looks like this is a Yamaha quirk; in this case CC 0-127 sends out squashed or expanded values. I will turn that function I posted into a more generalised function. It was only used once with the sequencer in the PLG150AN card.

Thanks Damien for confirming that I will have to do the hard yards and sniff out each message. It will be a challenge but now I know it is what I have to do. Just doing a JSON file save could work though, but it would be nice to access the system parameters and the patch name and reflect on the panel.

Regards,

**Reply by dnaldoog** [reply]
> I'll get back at you about the sysex for the Rev500. Took a quick look when you sent me the PDF the other day. Noticed an asterix *2 that you had to ask the Yamaha helpdesk for something. I couldn't figure out if it was about the parameters you are talking about. So, patience :).

I thought you were the helpdesk! LoL

**Comment by dnaldoog** [comment]
Excellent! Thanks for that. I’ll try it out.

**Reply by Ted-juh** [reply]
What I can't figure out is this: Is a timer used to let actions "after" a timer ran its course or can I also use it to run an action "during" the timer runs it's course?

**Comment by dnaldoog** [comment]
I wrote a _Class_ **Interpolate** that converts values 0-127 on a _uiSlider_ to the interpolated value. For example, you send 0-127 to (in my case the Yamaha REV500) but the display on the REV 500 might show as much as 200 values. This class updates _uiLabel_ in a layer sitting above _uiSlider_ to reflect the interpolated value.

This is just a rough template and would need to be adapted to a real world situation.  The range to interpolate is listed in _modulatorCustomIndex_ as a comma delimited pair of values i.e. 0,200 will interpolate CC 0-127 into 200 values to display on the panel.


----
![Interpolate](https://user-images.githubusercontent.com/76831683/118437424-94019b80-b71d-11eb-919d-44a8c20cedc0.png)

----
<pre>
<code>
Interpolate = {}
Interpolate.__index = Interpolate
setmetatable(
    Interpolate,
    {
        __call = function(cls, ...)
            local self = setmetatable({}, cls)
            self:_init(...)
            return self
        end
    }
)
------------------------------------------------------
function Interpolate:_init()
    setmetatable(Interpolate, {__index = Stack})
    self.mod = nil
    self.splitTab = {}
    self.midiMessage = 0
    return self
end -- method

function Interpolate:convertModValue(mod)
    self.mod = mod
    self:printValueToPanel()
    self:sendMidi()
    self:debug() -- delete or uncomment
end

function Interpolate:printValueToPanel()
    local l = self:getLabel()
    panel:getLabel(l):setText(tostring(self:getInterpolateValues()))
end

function Interpolate:getLabel()
    local mName = L(self.mod:getName())
    return mName:sub(1, 1, mName) .. "LABEL"
end

function Interpolate:getInterpolateValues()
    self.splitTab = self:split()
    local printTab = table.concat(self.splitTab, " ")
    local new_value = 0
    local old_value = self.mod:getModulatorValue()
    local old_min = 0
    local old_max = 127
    local new_min = self.splitTab[1]
    local new_max = self.splitTab[2]
    new_value = ((old_value - old_min) / (old_max - old_min)) * (new_max - new_min) + new_min
    return math.ceil(new_value)
end

function Interpolate:sendMidi()
    local channelOut = panel:getProperty("panelMidiOutputChannelDevice")
    local b0 = 0xb0 + (channelOut - 1) -- use this code for dynamic channel changes
    local b1 = self.mod:getMidiMessage():getProperty("midiMessageCtrlrNumber")
    local b2 = self.mod:getModulatorValue()
    self.midiMessage = sf("%.2X %.2X %.2X", b0, b1, b2)
    panel:sendMidiMessageNow(CtrlrMidiMessage(self.midiMessage))
end

function Interpolate:split()
    local source = self.mod:getProperty("modulatorCustomIndex")
    --console("str\160"..source)
    local elements = {}
    local pattern = "([^,]+)"
    string.gsub(
        source,
        pattern,
        function(value)
            elements[#elements + 1] = value
        end
    )
    return elements
end

function Interpolate:debug()
    if tog_debugging:getToggleState() then
        local nv = tostring(self:getInterpolateValues())
        panel:getLabel("d"):setText(
            sf("range=%s\nmodulatorValue=%s\n", self.mod:getValue(), table.concat(self.splitTab, ","))
        )
        panel:getLabel("d"):append(sf("interpolated value=%s", nv))
        panel:getLabel("MIDIOUT"):setText(sf("%s", self.midiMessage))
    else
    end
end
</code>
</pre>

[Interpolate Class_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/6491342/Interpolate.Class_1_0.zip)
